### PR TITLE
Fix NRE in primary constructor XML doc binder

### PR DESCRIPTION
**Pull request by jjonescz**
Fixes [AB#1931501](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1931501).

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit2/Semantics/PrimaryConstructorTests.cs [hunk 1]
```diff
@@ -6065,6 +6065,40 @@ public void XmlDoc([CombinatorialValues("class ", "struct")] string keyword)
             Assert.Equal("", constructor.GetParameters()[0].GetDocumentationCommentXml());
         }
 
+        [Theory, CombinatorialData, WorkItem("https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1931501")]
+        public void XmlDoc_InsideType(
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit2/Semantics/PrimaryConstructorTests.cs:6069]: >XmlDoc_InsideType [](http://example.com/codeflow?start=20&length=17)

It looks like the problem also affects records. Consider adding similar tests to record specific test files

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/PrimaryConstructorTests.cs:6069]: I think class and struct records are tested in different files

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 2]
```diff
@@ -1225,12 +1225,15 @@ private Binder GetParameterNameAttributeValueBinder(MemberDeclarationSyntax memb
                     _ = typeDeclaration.ParameterList;
                     Binder outerBinder = VisitCore(memberSyntax);
                     SourceNamedTypeSymbol type = ((NamespaceOrTypeSymbol)outerBinder.ContainingMemberOrLambda).GetSourceTypeMember((TypeDeclarationSyntax)memberSyntax);
-                    var primaryConstructor = type.PrimaryConstructor;
-
-                    if (primaryConstructor.SyntaxRef.SyntaxTree == memberSyntax.SyntaxTree &&
-                        primaryConstructor.GetSyntax() == memberSyntax)
+                    if (type is not null)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:1228]: > if (type is not null) [](http://example.com/codeflow?start=19&length=22)

While this change, perhaps, eliminates the crash for the specific scenario, it looks to me that ```GetAssociatedMemberForXmlSyntax``` shouldn't be returning type syntax in this case. This is not a doc comment on the type declaration. The binder thinks we are inside the type `C` and we trying to locate declaration of the same `C` within it. The search, of course, fails. It is quite possible that other code paths (other XML tags) in ```VisitXmlNameAttribute``` are negatively affected by the ```GetAssociatedMemberForXmlSyntax``` behavior as well.
To summarize, I think the change is fixing a symptom rather thana root cause of the problem.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 2 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: It looks like some other scenarios rely on this behavior. For example:

[https://github.com/dotnet/roslyn/blob/40d3f69e6d2e2d021c2ad546734ec6fb8d7c82d8/src/Compilers/CSharp/Test/Symbol/DocumentationComments/ParameterTests.cs#L760-L833](https://github.com/dotnet/roslyn/blob/40d3f69e6d2e2d021c2ad546734ec6fb8d7c82d8/src/Compilers/CSharp/Test/Symbol/DocumentationComments/ParameterTests.cs#L760-L833)

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: > It looks like some other scenarios rely on this behavior. 

Could you please elaborate? It doesn't look like the test is using primary constructors in any form and this code path is specific to primary constructors.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 3]
```diff
@@ -1401,7 +1401,12 @@ internal static MemberDeclarationSyntax GetAssociatedMemberForXmlSyntax(CSharpSy
                 MemberDeclarationSyntax memberSyntax = curr as MemberDeclarationSyntax;
                 if (memberSyntax != null)
                 {
-                    // CONSIDER: require that the xml syntax precede the start of the member span?
+                    // Require that the xml syntax precedes the start of the member span.
+                    if (memberSyntax.SpanStart < xmlSyntax.SpanStart)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:1405]: >if (memberSyntax.SpanStart < xmlSyntax.SpanStart) [](http://example.com/codeflow?start=20&length=49)

Perhaps the check will be more robust an intuitive if we check whether ```structuredTrivia``` is among ```memberSyntax.GetLeadingTrivia()``` nodes. At least it looks like this is how ```GetDocumentationCommentTriviaFromSyntaxNode``` helper locates doc comments

**Comment by AlekseyTs** [comment]
It looks like there are legitimate test failures, we probably need to adjust base-lines for tests that cover error scenarios.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 2 above)
Comment (jjonescz) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs]: I was referring to the behavior I tried to implement which you described in the comment above:

> it looks to me that `GetAssociatedMemberForXmlSyntax` shouldn't be returning type syntax in this case. This is not a doc comment on the type declaration. The binder thinks we are inside the type `C` and we trying to locate declaration of the same `C` within it

That's not specific to primary constructors, right?

**Review by cston — APPROVED** [review]
